keps/0030-immutable-parameters.md

```diff
@@ -0,0 +1,74 @@
+---
+kep-number: 30
+short-desc: Immutable parameters that can only be set on installation
+title: Immutable parameters
+authors:
+  - @aneumann82
+owners:
+  - @aneumann82
+editor:
+creation-date: 2020-04-28
+last-updated: 2020-04-28
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+## Summary
+
+This KEP describes a feature to allow parameters to be defined as immutable. Immutable parameters can not be updated after the installation of the operator
+
+## Motivation
+
+Certain aspects of an operator must be defined while installing an operator and can not be changed later on. 
+
+### Goals
+
+Make it possible for an operator developer to specify that a parameter can not be changed after the installation
+
+### Non-Goals
+
+- Parameters that can be set once after the installation - this KEP requires a parameter to be set in the installation process.
+
+## Proposal
+
+An flag `immutable` is added to the parameter definition. If that flag is true, the parameter needs to be set (or have a default) when the operator is installed. The parameter can not be updated with `kudo update -p` anymore.
+
+```yaml
+parameters:
+  - name: NUM_TOKENS
+    description: The number of tokens on each node of the cluster
+    default: 256
+    immutable: true
+```
+
+The default value for `immutable` is `false`.
+
+### User Stories
+
+- [#652](https://github.com/kudobuilder/kudo/issues/652) Make it possible to specify immutable parameter
+
+- NUM_TOKENS in the Cassandra operator
+- STORAGE_CLASS in the Zookeeper operator
+- DISK_SIZE, STORAGE_CLASS and PERSISTENT_STORAGE in the Kafka operator
+
+### Implementation Details/Notes/Constraints
+
+The KUDO CLI will not check if a parameter is immutable to allow providing an immutable parameter on the CLI or in a parameter file without changing it's value. This allows the user to keep the full operator configuration in a parameter file and use that for installation and updates.  
```

Comment:
Well, the alternative is that the KUDO CLI checks that a parameter is immutable independently of the value. If the CLI detects that `kudo update` is called with an immutable parameter it would error out.

The alternative is that immutable parameters can be specified with `kudo update` but their value will be silently ignored, which goes against "Fail loud and fast" principle

Reply:
I don't see the alt approach (last paragraph) as a violation of "fail load and fast".  I'm huge fan of "fail load and fast"... but to me that doesn't mean punish the user or reduce UX.  It means when the current experience is outside of the defined experience and expectations than fail... don't assume / hide.   We should design a good and convenient user experience.

Reply:
in the example of params: A, B and C, where B is immutable.. it should be ok for an update to have params for all 3 (as long as B hasn't changed values)... but it should also be ok for an update where A and C are specified (which implicitly means there is no change on B).

I believe this now states this is possible ... but we should accept the grammar change suggestion.